<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Note:-This-is-a-sample-solution-for-the-project.-Projects-will-NOT-be-graded-on-the-basis-of-how-well-the-submission-matches-this-sample-solution.-Projects-will-be-graded-on-the-basis-of-the-rubric-only." data-toc-modified-id="Note:-This-is-a-sample-solution-for-the-project.-Projects-will-NOT-be-graded-on-the-basis-of-how-well-the-submission-matches-this-sample-solution.-Projects-will-be-graded-on-the-basis-of-the-rubric-only.-1">Note: This is a sample solution for the project. Projects will NOT be graded on the basis of how well the submission matches this sample solution. Projects will be graded on the basis of the rubric only.</a></span></li><li><span><a href="#Context" data-toc-modified-id="Context-2">Context</a></span></li><li><span><a href="#Objective" data-toc-modified-id="Objective-3">Objective</a></span></li><li><span><a href="#Data-Description" data-toc-modified-id="Data-Description-4">Data Description</a></span></li><li><span><a href="#Let's-start-coding!" data-toc-modified-id="Let's-start-coding!-5">Let's start coding!</a></span><ul class="toc-item"><li><span><a href="#Importing-necessary-libraries" data-toc-modified-id="Importing-necessary-libraries-5.1">Importing necessary libraries</a></span></li></ul></li><li><span><a href="#Data-Preprocessing" data-toc-modified-id="Data-Preprocessing-6">Data Preprocessing</a></span><ul class="toc-item"><li><span><a href="#Processing-columns" data-toc-modified-id="Processing-columns-6.1">Processing columns</a></span></li><li><span><a href="#Feature-Engineering" data-toc-modified-id="Feature-Engineering-6.2">Feature Engineering</a></span><ul class="toc-item"><li><span><a href="#1.-Car-Brand-Name" data-toc-modified-id="1.-Car-Brand-Name-6.2.1">1. Car Brand Name</a></span></li><li><span><a href="#2.-Car-Model-Name" data-toc-modified-id="2.-Car-Model-Name-6.2.2">2. Car Model Name</a></span></li></ul></li></ul></li><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-7">Exploratory Data Analysis</a></span><ul class="toc-item"><li><span><a href="#Univariate-Analysis" data-toc-modified-id="Univariate-Analysis-7.1">Univariate Analysis</a></span></li><li><span><a href="#Kilometers_Driven" data-toc-modified-id="Kilometers_Driven-7.2"><em>Kilometers_Driven</em></a></span></li><li><span><a href="#Bivariate-Distributions" data-toc-modified-id="Bivariate-Distributions-7.3">Bivariate Distributions</a></span></li></ul></li><li><span><a href="#Data-Preprocessing-(contd.)" data-toc-modified-id="Data-Preprocessing-(contd.)-8">Data Preprocessing (contd.)</a></span><ul class="toc-item"><li><span><a href="#Missing-Value-Treatment" data-toc-modified-id="Missing-Value-Treatment-8.1">Missing Value Treatment</a></span></li></ul></li><li><span><a href="#Linear-Model-Building" data-toc-modified-id="Linear-Model-Building-9">Linear Model Building</a></span><ul class="toc-item"><li><span><a href="#Model-Performance-Check" data-toc-modified-id="Model-Performance-Check-9.1">Model Performance Check</a></span></li><li><span><a href="#Forward-Feature-Selection" data-toc-modified-id="Forward-Feature-Selection-9.2">Forward Feature Selection</a></span></li></ul></li><li><span><a href="#Observations" data-toc-modified-id="Observations-10">Observations</a></span></li><li><span><a href="#Business-Insights-and-Recommendations" data-toc-modified-id="Business-Insights-and-Recommendations-11">Business Insights and Recommendations</a></span></li><li><span><a href="#Add-on:-Analysing-predictions-where-we-were-way-off-the-mark" data-toc-modified-id="Add-on:-Analysing-predictions-where-we-were-way-off-the-mark-12">Add-on: Analysing predictions where we were way off the mark</a></span></li><li><span><a href="#Add-on:-Advanced-Data-Preprocessing" data-toc-modified-id="Add-on:-Advanced-Data-Preprocessing-13">Add-on: Advanced Data Preprocessing</a></span></li></ul></div>

## Note: This is a sample solution for the project. Projects will NOT be graded on the basis of how well the submission matches this sample solution. Projects will be graded on the basis of the rubric only.

## Context

There is a huge demand for used cars in the Indian Market today. As sales of new cars have slowed down in the recent past, the pre-owned car market has continued to grow over the past years and is larger than the new car market now. Cars4U is a budding tech start-up that aims to find footholes in this market.

In 2018-19, while new car sales were recorded at 3.6 million units, around 4 million second-hand cars were bought and sold. There is a slowdown in new car sales and that could mean that the demand is shifting towards the pre-owned market. In fact, some car sellers replace their old cars with pre-owned cars instead of buying new ones. Unlike new cars, where price and supply are fairly deterministic and managed by OEMs (Original Equipment Manufacturer / except for dealership level discounts which come into play only in the last stage of the customer journey), used cars are very different beasts with huge uncertainty in both pricing and supply. Keeping this in mind, the pricing scheme of these used cars becomes important in order to grow in the market.

As a senior data scientist at Cars4U, you have to come up with a pricing model that can effectively predict the price of used cars and can help the business in devising profitable strategies using differential pricing. For example, if the business knows the market price, it will never sell anything below it. 


## Objective

To explore and visualize the dataset, build a linear regression model to predict the prices of used cars, and generate a set of insights and recommendations that will help the business.


## Data Description

The data contains the different attributes of used cars sold in different locations. The detailed data dictionary is given below. 

**Data Dictionary** 

- S.No.: Serial number
- Name: Name of the car which includes brand name and model name
- Location: Location in which the car is being sold or is available for purchase (cities)
- Year: Manufacturing year of the car
- Kilometers_driven: The total kilometers driven in the car by the previous owner(s) in km
- Fuel_Type: The type of fuel used by the car (Petrol, Diesel, Electric, CNG, LPG)
- Transmission: The type of transmission used by the car (Automatic/Manual)
- Owner: Type of ownership
- Mileage: The standard mileage offered by the car company in kmpl or km/kg
- Engine: The displacement volume of the engine in CC
- Power: The maximum power of the engine in bhp
- Seats: The number of seats in the car
- New_Price: The price of a new car of the same model in INR Lakhs (1 Lakh = 100,000 INR)
- Price: The price of the used car in INR Lakhs

## Let's start coding!

### Importing necessary libraries

In [36]:
# this will help in making the Python code more structured automatically (good coding practice)
%load_ext nb_black

# Libraries to help with reading and manipulating data
import numpy as np
import pandas as pd

# Libraries to help with data visualization
import matplotlib.pyplot as plt
import seaborn as sns

sns.set()

# Removes the limit for the number of displayed columns
pd.set_option("display.max_columns", None)
# Sets the limit for the number of displayed rows
pd.set_option("display.max_rows", 500)

# to split the data into train and test
from sklearn.model_selection import train_test_split

# to build linear regression_model
from sklearn.linear_model import LinearRegression

# to check model performance
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# to suppress warnings
import warnings

warnings.filterwarnings("ignore")

The nb_black extension is already loaded. To reload it, use:
  %reload_ext nb_black


<IPython.core.display.Javascript object>

In [37]:
# loading the dataset
data = pd.read_csv("./datasets/used_cars_data.csv")

<IPython.core.display.Javascript object>

In [38]:
# checking shape of the data
print(f"There are {data.shape[0]} rows and {data.shape[1]} columns.")

There are 7253 rows and 14 columns.


<IPython.core.display.Javascript object>

In [39]:
# let's view a sample of the data
data.sample(n=10, random_state=1)

S.No.                                       Name    Location  Year  \
2397   2397             Ford EcoSport 1.5 Petrol Trend     Kolkata  2016   
3777   3777                     Maruti Wagon R VXI 1.2       Kochi  2015   
4425   4425                     Ford Endeavour 4x2 XLT   Hyderabad  2007   
3661   3661  Mercedes-Benz E-Class E250 CDI Avantgrade  Coimbatore  2016   
4514   4514       Hyundai Xcent 1.2 Kappa AT SX Option       Kochi  2016   
599     599             Toyota Innova Crysta 2.8 ZX AT  Coimbatore  2019   
186     186  Mercedes-Benz E-Class E250 CDI Avantgrade   Bangalore  2014   
305     305     Audi A6 2011-2015 2.0 TDI Premium Plus       Kochi  2014   
4582   4582                      Hyundai i20 1.2 Magna     Kolkata  2011   
5434   5434           Honda WR-V Edge Edition i-VTEC S       Kochi  2019   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
2397              21460    Petrol       Manual      First   17.0 kmpl   
3777              49818    Petrol       Manual      First   21.5 kmpl   
4425             130000    Diesel       Manual      First   13.1 kmpl   
3661              39753    Diesel    Automatic      First   13.0 kmpl   
4514              45560    Petrol    Automatic      First   16.9 kmpl   
599               40674    Diesel    Automatic      First  11.36 kmpl   
186               37382    Diesel    Automatic      First   13.0 kmpl   
305               61726    Diesel    Automatic      First  17.68 kmpl   
4582              36000    Petrol       Manual      First   18.5 kmpl   
5434              13913    Petrol       Manual      First   17.5 kmpl   

       Engine       Power  Seats   New_Price  Price  
2397  1497 CC  121.36 bhp    5.0   9.47 Lakh   6.00  
3777  1197 CC   81.80 bhp    5.0   5.44 Lakh   4.11  
4425  2499 CC     141 bhp    7.0         NaN   6.00  
3661  2143 CC   201.1 bhp    5.0         NaN  35.28  
4514  1197 CC      82 bhp    5.0         NaN   6.34  
599   2755 CC   171.5 bhp    7.0  28.05 Lakh  24.82  
186   2143 CC   201.1 bhp    5.0         NaN  32.00  
305   1968 CC  174.33 bhp    5.0         NaN  20.77  
4582  1197 CC      80 bhp    5.0         NaN   2.50  
5434  1199 CC    88.7 bhp    5.0   9.36 Lakh   8.20

<IPython.core.display.Javascript object>

**Observations**

- `S.No.` is just an index for the data entry and will add no value to our analysis. So, we will drop it.

- `Name` contains a lot of model information. Let us check how many individual names we have. If they are too many, we can process this column to extract important information.

- `Mileage`, `Engine`, and `Power` columns will also need some processing before we are able to explore them. We'll have to extract numerical information from these columns.

- The `New_Price` column also needs some processing. This one also contains strings and a lot of missing values.

In [40]:
# let's create a copy of the data to avoid any changes to original data
df = data.copy()

<IPython.core.display.Javascript object>

In [41]:
# drop the S.No. column as it does not add any value to the analysis
df.drop("S.No.", axis=1, inplace=True)

<IPython.core.display.Javascript object>

In [42]:
# checking column datatypes and number of non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7253 entries, 0 to 7252
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               7253 non-null   object 
 1   Location           7253 non-null   object 
 2   Year               7253 non-null   int64  
 3   Kilometers_Driven  7253 non-null   int64  
 4   Fuel_Type          7253 non-null   object 
 5   Transmission       7253 non-null   object 
 6   Owner_Type         7253 non-null   object 
 7   Mileage            7251 non-null   object 
 8   Engine             7207 non-null   object 
 9   Power              7078 non-null   object 
 10  Seats              7200 non-null   float64
 11  New_Price          1006 non-null   object 
 12  Price              6019 non-null   float64
dtypes: float64(2), int64(2), object(9)
memory usage: 736.8+ KB


<IPython.core.display.Javascript object>

In [43]:
df.isnull().sum()

Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 46
Power                 175
Seats                  53
New_Price            6247
Price                1234
dtype: int64

<IPython.core.display.Javascript object>

**Observations**

- `Mileage`, `Engine`, `Power`, and `New_Price` are *object* type columns when they should ideally be numerical.
- To be able to get summary statistics for these columns, we will have to process them first.

In [44]:
# checking for duplicate values
df.duplicated().sum()

1

<IPython.core.display.Javascript object>

- There is one duplicate value in the data.
- Let's take a closer look at it.

In [45]:
df[df.duplicated(keep=False) == True]

Name Location  Year  Kilometers_Driven Fuel_Type  \
6498  Honda City 1.5 E MT   Mumbai  2010              52000    Petrol   
6582  Honda City 1.5 E MT   Mumbai  2010              52000    Petrol   

     Transmission Owner_Type    Mileage   Engine    Power  Seats New_Price  \
6498       Manual      First  17.0 kmpl  1497 CC  118 bhp    5.0       NaN   
6582       Manual      First  17.0 kmpl  1497 CC  118 bhp    5.0       NaN   

      Price  
6498    NaN  
6582    NaN

<IPython.core.display.Javascript object>

**Observations**

- There is a good chance that two cars of the same build were sold in the same location.
- But it is highly unlikely that both of them will have the same number of kilometers driven.
- So, we will drop the row which occurs second.

In [46]:
df.drop(6582, inplace=True)

<IPython.core.display.Javascript object>

## Data Preprocessing

### Processing columns

- Let's process the `Mileage`, `Engine`, `Power`, and `New_Price` columns to extract numerical values from them.

**1. `Mileage`**

* We have car mileage in two units as per the data dictionary - kmpl and km/kg.

* After quick research on the internet, it is clear that these 2 units are used for cars of 2 different fuel types.

    * kmpl (kilometers per liter) is used for petrol and diesel cars.
    * km/kg (kilometers per kg) is used for CNG and LPG-based engines.

* We have the variable `Fuel_type` in our data. We will check if these observations hold true in our data too.

In [47]:
# viewing the column values
df["Mileage"].head(10)

0    26.6 km/kg
1    19.67 kmpl
2     18.2 kmpl
3    20.77 kmpl
4     15.2 kmpl
5    21.1 km/kg
6    23.08 kmpl
7    11.36 kmpl
8    20.54 kmpl
9     22.3 kmpl
Name: Mileage, dtype: object

<IPython.core.display.Javascript object>

In [48]:
df_mileage = df["Mileage"].str.split(" ", expand=True)
df_mileage.head()

0      1
0   26.6  km/kg
1  19.67   kmpl
2   18.2   kmpl
3  20.77   kmpl
4   15.2   kmpl

<IPython.core.display.Javascript object>

In [49]:
# let's verify that there are two units
df_mileage[1].value_counts()

kmpl     7176
km/kg      74
Name: 1, dtype: int64

<IPython.core.display.Javascript object>

In [50]:
# we will create two new columns for mileage values and units
df["mileage_num"] = df_mileage[0].astype(float)
df["mileage_unit"] = df_mileage[1]

# Checking the new dataframe
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  mileage_num mileage_unit  
0        NaN   1.75        26.60        km/kg  
1        NaN  12.50        19.67         kmpl  
2  8.61 Lakh   4.50        18.20         kmpl  
3        NaN   6.00        20.77         kmpl  
4        NaN  17.74        15.20         kmpl

<IPython.core.display.Javascript object>

In [51]:
# Let's check if the units correspond to the fuel types
df.groupby(by=["Fuel_Type", "mileage_unit"]).size()

Fuel_Type  mileage_unit
CNG        km/kg             62
Diesel     kmpl            3852
LPG        km/kg             12
Petrol     kmpl            3324
dtype: int64

<IPython.core.display.Javascript object>

* As expected, km/kg is for CNG/LPG cars and kmpl is for Petrol and Diesel cars.

**2. `Engine`**

* As per the data dictionary, the `Engine` column indicates the displacement volume of the engine in CC.
* Let's extract the numerical part of the column values.

In [52]:
# viewing the column values
df["Engine"].head(10)

0     998 CC
1    1582 CC
2    1199 CC
3    1248 CC
4    1968 CC
5     814 CC
6    1461 CC
7    2755 CC
8    1598 CC
9    1248 CC
Name: Engine, dtype: object

<IPython.core.display.Javascript object>

- The numeric values and units are separated by a space. We will split the values by space.

In [53]:
df_engine = df["Engine"].str.split(" ", expand=True)
df_engine.head()

0   1
0   998  CC
1  1582  CC
2  1199  CC
3  1248  CC
4  1968  CC

<IPython.core.display.Javascript object>

In [54]:
# let's verify that there is only one unit
df_engine[1].value_counts()

CC    7206
Name: 1, dtype: int64

<IPython.core.display.Javascript object>

In [55]:
# we will create a new column for engine values
df["engine_num"] = df_engine[0].astype(float)

# Checking the new dataframe
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  mileage_num mileage_unit  engine_num  
0        NaN   1.75        26.60        km/kg       998.0  
1        NaN  12.50        19.67         kmpl      1582.0  
2  8.61 Lakh   4.50        18.20         kmpl      1199.0  
3        NaN   6.00        20.77         kmpl      1248.0  
4        NaN  17.74        15.20         kmpl      1968.0

<IPython.core.display.Javascript object>

**3. `Power`**

* As per the data dictionary, the `Power` column indicates the maximum power of the engine in bhp.
* Let's extract the numerical part of the column values.

In [56]:
# viewing the column values
df["Power"].head(10)

0    58.16 bhp
1    126.2 bhp
2     88.7 bhp
3    88.76 bhp
4    140.8 bhp
5     55.2 bhp
6     63.1 bhp
7    171.5 bhp
8    103.6 bhp
9       74 bhp
Name: Power, dtype: object

<IPython.core.display.Javascript object>

In [57]:
df_power = df["Power"].str.split(" ", expand=True)
df_power.head()

0    1
0  58.16  bhp
1  126.2  bhp
2   88.7  bhp
3  88.76  bhp
4  140.8  bhp

<IPython.core.display.Javascript object>

In [58]:
# let's verify that there is only one unit
df_power[1].value_counts()

bhp    7077
Name: 1, dtype: int64

<IPython.core.display.Javascript object>

In [59]:
# we will create a new column for power values
df["power_num"] = df_power[0].astype(float)

# Checking the new dataframe
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  mileage_num mileage_unit  engine_num  power_num  
0        NaN   1.75        26.60        km/kg       998.0      58.16  
1        NaN  12.50        19.67         kmpl      1582.0     126.20  
2  8.61 Lakh   4.50        18.20         kmpl      1199.0      88.70  
3        NaN   6.00        20.77         kmpl      1248.0      88.76  
4        NaN  17.74        15.20         kmpl      1968.0     140.80

<IPython.core.display.Javascript object>

**4. `New_Price`**

* We know that `New_Price` is the price of a new car of the same model in INR Lakhs (1 Lakh INR = 100, 000 INR)
* Let's extract the numerical part of the column values.

In [60]:
# viewing the column values
df["New_Price"].head(10)

0          NaN
1          NaN
2    8.61 Lakh
3          NaN
4          NaN
5          NaN
6          NaN
7      21 Lakh
8          NaN
9          NaN
Name: New_Price, dtype: object

<IPython.core.display.Javascript object>

In [61]:
df_new_price = df["New_Price"].str.split(" ", expand=True)
df_new_price.head()

0     1
0   NaN   NaN
1   NaN   NaN
2  8.61  Lakh
3   NaN   NaN
4   NaN   NaN

<IPython.core.display.Javascript object>

In [62]:
# let's verify that there is only one unit
df_new_price[1].value_counts()

Lakh    986
Cr       20
Name: 1, dtype: int64

<IPython.core.display.Javascript object>

* There are two units - Lakh and Cr.
* Let us convert all values to lakhs using the conversion rate 1 Cr = 100 Lakhs.

In [63]:
# let's define a function to convert the New_Price column to float


def new_price_to_num(new_price_val):
    """
    This function takes in a string representing a new car price
    and converts it to a number. For example, '8.61 Lakh' becomes 8.61.
    If the input is already numeric, which probably means it's NaN,
    this function just returns np.nan.
    """
    if isinstance(new_price_val, str):  # checks if `new_price_val` is a string
        multiplier = 1  # handles Lakh vs Cr values
        if new_price_val.endswith("Lakh"):
            multiplier = 1
        elif new_price_val.endswith("Cr"):
            multiplier = 100
        return float(new_price_val.replace(" Lakh", "").replace(" Cr", "")) * multiplier
    else:  # this happens when the current new_price is np.nan
        return np.nan

<IPython.core.display.Javascript object>

In [64]:
# let's apply the function to the New_Price column
df["new_price_num"] = df["New_Price"].apply(new_price_to_num)

# Checking the new dataframe
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  mileage_num mileage_unit  engine_num  power_num  \
0        NaN   1.75        26.60        km/kg       998.0      58.16   
1        NaN  12.50        19.67         kmpl      1582.0     126.20   
2  8.61 Lakh   4.50        18.20         kmpl      1199.0      88.70   
3        NaN   6.00        20.77         kmpl      1248.0      88.76   
4        NaN  17.74        15.20         kmpl      1968.0     140.80   

   new_price_num  
0            NaN  
1            NaN  
2           8.61  
3            NaN  
4            NaN

<IPython.core.display.Javascript object>

- We have extracted numerical values from the `Mileage`, `Engine`, `Power`, and `New_Price` columns.

### Feature Engineering

- The `Name` column in the current format might not be very useful in our analysis.
- Since the name contains both the brand name and the model name of the vehicle, the column would have too many unique values to be useful in prediction.

In [65]:
# checking number of unique values
df["Name"].nunique()

2041

<IPython.core.display.Javascript object>

- With 2041 unique names, car names are not going to be great predictors of the price in our current data.

- But we can process this column to extract important information and see if that reduces the number of levels for this information.

#### 1. Car Brand Name

In [70]:
# extracting brand names
df["Brand"] = df["Name"].apply(lambda x: x.split(" ")[0].lower())
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  mileage_num mileage_unit  engine_num  power_num  \
0        NaN   1.75        26.60        km/kg       998.0      58.16   
1        NaN  12.50        19.67         kmpl      1582.0     126.20   
2  8.61 Lakh   4.50        18.20         kmpl      1199.0      88.70   
3        NaN   6.00        20.77         kmpl      1248.0      88.76   
4        NaN  17.74        15.20         kmpl      1968.0     140.80   

   new_price_num    Brand  
0            NaN   maruti  
1            NaN  hyundai  
2           8.61    honda  
3            NaN   maruti  
4            NaN     audi

<IPython.core.display.Javascript object>

In [71]:
# checking the unique values and their number of occurences
df["Brand"].value_counts()

maruti           1444
hyundai          1340
honda             742
toyota            507
mercedes-benz     380
volkswagen        374
ford              351
mahindra          331
bmw               312
audi              285
tata              228
skoda             202
renault           170
chevrolet         151
nissan            117
land               67
jaguar             48
fiat               38
mitsubishi         36
mini               31
volvo              28
jeep               19
porsche            19
datsun             17
isuzu               5
force               3
bentley             2
lamborghini         1
opelcorsa           1
smart               1
ambassador          1
hindustan           1
Name: Brand, dtype: int64

<IPython.core.display.Javascript object>

- There are 32 brands of cars in the data.

#### 2. Car Model Name

In [ ]:
# extracting model names
df["Model"] = df["Name"].apply(lambda x: x.split(" ")[1].lower())
df.head()

In [ ]:
# checking the unique values and their number of occurences
df["Model"].value_counts()

- There are 218 different car models in the data.

**Before we start exploring the data further, let's quickly check the missing values in the data.**

In [ ]:
df.isnull().sum()

**Observations**

* 2 cars don't have entries for Mileage.

* Engine displacement information of 46 observations is missing, the maximum power of 175 entries is missing, and information about the number of seats is not available for 53 entries.

* The new price is missing for 6247 cars in the dataset. We'll have to see if there is a pattern here to determine if we can impute these or if we should drop this column altogether.

* Price is also missing for 1234 entries. Since price is the response variable that we want to predict, we will have to drop these rows when we actually build a model.
    - These rows will not be able to help us in modeling or model evaluation. But while we are analyzing the distributions and doing missing value imputations, we will keep using information from these rows.

In [ ]:
# dropping the redundant columns.
df.drop(columns=["Mileage", "Engine", "Power", "New_Price"], inplace=True)

## Exploratory Data Analysis

**Let's check the statistical summary of the numerical variables.**

In [67]:
df.describe().T

count          mean           std      min        25%  \
Year               7252.0   2013.365830      3.254405  1996.00   2011.000   
Kilometers_Driven  7252.0  58699.986900  84433.505524   171.00  34000.000   
Seats              7199.0      5.279761      0.811709     0.00      5.000   
Price              6019.0      9.479468     11.187917     0.44      3.500   
mileage_num        7250.0     18.141738      4.562492     0.00     15.170   
engine_num         7206.0   1616.590064    595.324779    72.00   1198.000   
power_num          7077.0    112.764474     53.497297    34.20     75.000   
new_price_num      1006.0     22.779692     27.759344     3.91      7.885   

                        50%         75%         max  
Year                2014.00   2016.0000     2019.00  
Kilometers_Driven  53429.00  73000.0000  6500000.00  
Seats                  5.00      5.0000       10.00  
Price                  5.64      9.9500      160.00  
mileage_num           18.16     21.1000       33.54  
engine_num          1493.00   1968.0000     5998.00  
power_num             94.00    138.1000      616.00  
new_price_num         11.57     26.0425      375.00

<IPython.core.display.Javascript object>

**Observations**

1. `Kilometers_Driven` values have an incredibly high range. We should check a few of the extreme values to get a sense of the data.

2. The minimum and the maximum number of seats in the car also warrant a quick check. On average, a car seems to have 5 seats, which is about right.

3. We have used cars being sold at less than a lakh INR and as high as 160 lakh INR, as we saw for the Lamborghini earlier. We might have to check for outliers to build a robust model.

4. The minimum mileage being 0 is also concerning. We'll have to check what is going on.

5. Engine and Power mean and median values are not very different. Only someone with more domain knowledge would be able to comment further on these attributes.

6. The new price range seems right. We have both budget-friendly Maruti cars and Lamborghinis in our stock. Mean being twice that of the median suggests that there are only a few very high range brands, which again makes sense.

In [68]:
# checking extreme values in Kilometers_Driven
df.sort_values(by=["Kilometers_Driven"], ascending=False).head(10)

Name   Location  Year  \
2328                        BMW X5 xDrive 30d M Sport    Chennai  2017   
340            Skoda Octavia Ambition Plus 2.0 TDI AT    Kolkata  2013   
1860                 Volkswagen Vento Diesel Highline    Chennai  2013   
358                             Hyundai i10 Magna 1.2    Chennai  2009   
2823  Volkswagen Jetta 2013-2015 2.0L TDI Highline AT    Chennai  2015   
3092                             Honda City i VTEC SV    Kolkata  2015   
4491                   Hyundai i20 Magna Optional 1.2  Bangalore  2013   
6921                      Maruti Swift Dzire Tour LDI     Jaipur  2012   
3649                                   Tata Indigo LS     Jaipur  2008   
1528      Toyota Innova 2.5 G (Diesel) 8 Seater BS IV  Hyderabad  2005   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
2328            6500000    Diesel    Automatic      First  15.97 kmpl   
340              775000    Diesel    Automatic      First   19.3 kmpl   
1860             720000    Diesel       Manual      First  20.54 kmpl   
358              620000    Petrol       Manual      First  20.36 kmpl   
2823             480000    Diesel    Automatic      First  16.96 kmpl   
3092             480000    Petrol       Manual      First   17.4 kmpl   
4491             445000    Petrol       Manual      First   18.5 kmpl   
6921             350000    Diesel       Manual      First   23.4 kmpl   
3649             300000    Diesel       Manual      First   17.0 kmpl   
1528             299322    Diesel       Manual      First   12.8 kmpl   

       Engine       Power  Seats New_Price  Price  mileage_num mileage_unit  \
2328  2993 CC     258 bhp    5.0       NaN  65.00        15.97         kmpl   
340   1968 CC     141 bhp    5.0       NaN   7.50        19.30         kmpl   
1860  1598 CC   103.6 bhp    5.0       NaN   5.90        20.54         kmpl   
358   1197 CC    78.9 bhp    5.0       NaN   2.70        20.36         kmpl   
2823  1968 CC  138.03 bhp    5.0       NaN  13.00        16.96         kmpl   
3092  1497 CC   117.3 bhp    5.0       NaN   5.00        17.40         kmpl   
4491  1197 CC    82.9 bhp    5.0       NaN   4.45        18.50         kmpl   
6921  1248 CC      74 bhp    5.0       NaN    NaN        23.40         kmpl   
3649  1405 CC      70 bhp    5.0       NaN   1.00        17.00         kmpl   
1528  2494 CC     102 bhp    8.0       NaN   4.00        12.80         kmpl   

      engine_num  power_num  new_price_num  
2328      2993.0     258.00            NaN  
340       1968.0     141.00            NaN  
1860      1598.0     103.60            NaN  
358       1197.0      78.90            NaN  
2823      1968.0     138.03            NaN  
3092      1497.0     117.30            NaN  
4491      1197.0      82.90            NaN  
6921      1248.0      74.00            NaN  
3649      1405.0      70.00            NaN  
1528      2494.0     102.00            NaN

<IPython.core.display.Javascript object>

**Observations**

* It looks like the first row here is a data entry error. A car manufactured as recently as 2017 having been driven 65,00,000 km is almost impossible.

* The other observations that follow are also on a higher end. There is a good chance that these are outliers. We'll look at this further while doing the univariate analysis.

In [69]:
# checking extreme values in Kilometers_Driven
df.sort_values(by=["Kilometers_Driven"], ascending=True).head(10)

Name   Location  Year  \
1361               Maruti Alto 800 2016-2019 VXI     Mumbai  2019   
5606                  Maruti Wagon R ZXI AMT 1.2     Mumbai  2019   
2054                      Maruti Ignis 1.2 Alpha     Mumbai  2017   
6201                             Maruti Alto LXI    Kolkata  2019   
1598                        Tata Xenon XT EX 4X2     Jaipur  2017   
173             Hyundai Grand i10 1.2 Kappa Asta    Kolkata  2019   
1198           Volkswagen Polo 1.0 MPI Trendline  Hyderabad  2019   
5941                            Renault KWID RXL       Pune  2018   
1242              Jaguar XE 2.0L Diesel Prestige      Delhi  2018   
2646  Ford Figo 2015-2019 1.2P Sports Edition MT     Mumbai  2019   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
1361                171    Petrol       Manual      First   24.7 kmpl   
5606                600    Petrol    Automatic      First   21.5 kmpl   
2054               1000    Petrol       Manual      First  20.89 kmpl   
6201               1000    Petrol       Manual      First   24.7 kmpl   
1598               1000    Diesel       Manual      First  13.49 kmpl   
173                1000    Petrol       Manual      First   18.9 kmpl   
1198               1000    Petrol       Manual      First  18.78 kmpl   
5941               1000    Petrol       Manual      First  25.17 kmpl   
1242               1000    Diesel    Automatic      First   13.6 kmpl   
2646               1000    Petrol       Manual      First  18.12 kmpl   

       Engine       Power  Seats   New_Price  Price  mileage_num mileage_unit  \
1361   796 CC    47.3 bhp    5.0         NaN   3.60        24.70         kmpl   
5606  1197 CC   81.80 bhp    5.0    6.8 Lakh   6.25        21.50         kmpl   
2054  1197 CC   81.80 bhp    5.0   7.87 Lakh   6.36        20.89         kmpl   
6201   796 CC    47.3 bhp    5.0   3.99 Lakh    NaN        24.70         kmpl   
1598  2179 CC  138.03 bhp    5.0         NaN   7.50        13.49         kmpl   
173   1197 CC   81.86 bhp    5.0   7.39 Lakh   6.50        18.90         kmpl   
1198   999 CC      75 bhp    5.0   6.74 Lakh   5.68        18.78         kmpl   
5941   799 CC    53.3 bhp    5.0   4.45 Lakh   3.40        25.17         kmpl   
1242  1999 CC     177 bhp    5.0  52.77 Lakh  40.00        13.60         kmpl   
2646  1196 CC    86.8 bhp    5.0         NaN   6.80        18.12         kmpl   

      engine_num  power_num  new_price_num  
1361       796.0      47.30            NaN  
5606      1197.0      81.80           6.80  
2054      1197.0      81.80           7.87  
6201       796.0      47.30           3.99  
1598      2179.0     138.03            NaN  
173       1197.0      81.86           7.39  
1198       999.0      75.00           6.74  
5941       799.0      53.30           4.45  
1242      1999.0     177.00          52.77  
2646      1196.0      86.80            NaN

<IPython.core.display.Javascript object>

**Observations**

* After looking at the columns `Year`, `New_Price`, and `Price`, these entries seem feasible.

* 1000 might be the default value in this case. Quite a few cars having driven exactly 1000 km is suspicious.

In [ ]:
# checking extreme values in Seats
df.sort_values(by=["Seats"], ascending=True).head(5)

**Observations**

* Audi A4 having 0 seats is clearly a data entry error.
* This column warrants some outlier treatment, or we can treat Seats = 0 as a missing value.
* Overall, there doesn't seem not much to be concerned about here.
* Let's check if we have a similar car in our dataset.

In [ ]:
df[df["Name"].str.startswith("Audi A4 3.2")]

* Looks like an Audi A4 typically has 5 seats.

In [ ]:
# Let us replace #seats in row index 3999 form 0 to 5
df.loc[3999, "Seats"] = 5.0

In [ ]:
# checking extreme values in Seats
df.sort_values(by=["Seats"], ascending=False).head(5)

* A Toyota Qualis does have 10 seats and so does a Tata Sumo. We don't see any data entry error here.

In [ ]:
# checking extreme values in Mileage
df.sort_values(by=["mileage_num"], ascending=True).head(10)

* We will have to treat Mileage = 0 as missing values

In [ ]:
# checking extreme values in Mileage
df.sort_values(by=["mileage_num"], ascending=False).head(10)

* Maruti Wagon R and Maruti Alto CNG versions are budget-friendly cars with high mileage, so these data points are fine.

In [ ]:
df.dtypes[df.dtypes == "object"].index

In [ ]:
# looking at value counts for non-numeric features

num_to_display = 10  # defining this up here so it's easy to change later

for colname in df.dtypes[df.dtypes == "object"].index:
    val_counts = df[colname].value_counts(dropna=False)  # will also show the NA counts
    print(val_counts[:num_to_display])

    if len(val_counts) > num_to_display:
        print(f"Only displaying first {num_to_display} of {len(val_counts)} values.")
    print("-" * 50, "\n")  # just for more space in between

- We had checked cars of different `Fuel_Type` earlier, but we did not encounter the 2 electric cars. Let us check why.

In [ ]:
df.loc[df["Fuel_Type"] == "Electric"]

**Observations**

* Mileage values for these cars are NaN, that's why we did not encounter these earlier with *groupby()*.

* Electric cars are very new in the market and very rare in our dataset. We can consider dropping these two observations if they turn out to be outliers later.

* There is a good chance that we will not be able to create a good price prediction model for electric cars with the currently available data.

In [ ]:
# checking used ar prices grouped by brand
df.groupby(["Brand"])["Price"].mean().sort_values(ascending=False)

**Observations**

* The output is very close to our expectation (domain knowledge) in terms of brand order.
* The mean price of a used Lamborghini is 120 Lakhs and that of cars from other luxury brands follow in descending order.
* Towards the bottom end, we have the more budget-friendly brands.

### Univariate Analysis

***Price***

In [ ]:
# function to plot a boxplot and a histogram along the same scale.


def histogram_boxplot(data, feature, figsize=(12, 7), kde=False, bins=None):
    """
    Boxplot and histogram combined

    data: dataframe
    feature: dataframe column
    figsize: size of figure (default (12,7))
    kde: whether to the show density curve (default False)
    bins: number of bins for histogram (default None)
    """
    f2, (ax_box2, ax_hist2) = plt.subplots(
        nrows=2,  # Number of rows of the subplot grid= 2
        sharex=True,  # x-axis will be shared among all subplots
        gridspec_kw={"height_ratios": (0.25, 0.75)},
        figsize=figsize,
    )  # creating the 2 subplots
    sns.boxplot(
        data=data, x=feature, ax=ax_box2, showmeans=True, color="violet"
    )  # boxplot will be created and a star will indicate the mean value of the column
    sns.histplot(
        data=data, x=feature, kde=kde, ax=ax_hist2, bins=bins, palette="winter"
    ) if bins else sns.histplot(
        data=data, x=feature, kde=kde, ax=ax_hist2
    )  # For histogram
    ax_hist2.axvline(
        data[feature].mean(), color="green", linestyle="--"
    )  # Add mean to the histogram
    ax_hist2.axvline(
        data[feature].median(), color="black", linestyle="-"
    )  # Add median to the histogram

In [ ]:
histogram_boxplot(df, "Price", kde=True)

**Observations**

* This is a highly skewed distribution.
* Let us use log transformation on this column to see if that helps normalize the distribution.

In [ ]:
# Creating a new column with the transformed variable
df["price_log"] = np.log(df["Price"])

In [ ]:
histogram_boxplot(df, "price_log", kde=True)

* Using the log transformation has definitely helped in reducing the skew.

### *Kilometers_Driven*

In [ ]:
histogram_boxplot(df, "Kilometers_Driven", bins=100, kde=True)

**Observations**

* This is another highly skewed distribution.
* Let us use log transformation on this column too.

In [ ]:
df["kilometers_driven_log"] = np.log(df["Kilometers_Driven"])

In [ ]:
histogram_boxplot(df, "kilometers_driven_log", bins=100, kde=True)

* Transformation has reduced the extreme skewness.

In [ ]:
# function to create labeled barplots


def labeled_barplot(data, feature, perc=False, n=None):
    """
    Barplot with percentage at the top

    data: dataframe
    feature: dataframe column
    perc: whether to display percentages instead of count (default is False)
    n: displays the top n category levels (default is None, i.e., display all levels)
    """

    total = len(data[feature])  # length of the column
    count = data[feature].nunique()
    if n is None:
        plt.figure(figsize=(count + 1, 5))
    else:
        plt.figure(figsize=(n + 1, 5))

    plt.xticks(rotation=90, fontsize=15)
    ax = sns.countplot(
        data=data,
        x=feature,
        palette="Paired",
        order=data[feature].value_counts().index[:n].sort_values(),
    )

    for p in ax.patches:
        if perc == True:
            label = "{:.1f}%".format(
                100 * p.get_height() / total
            )  # percentage of each class of the category
        else:
            label = p.get_height()  # count of each level of the category

        x = p.get_x() + p.get_width() / 2  # width of the plot
        y = p.get_height()  # height of the plot

        ax.annotate(
            label,
            (x, y),
            ha="center",
            va="center",
            size=12,
            xytext=(0, 5),
            textcoords="offset points",
        )  # annotate the percentage

    plt.show()  # show the plot

In [ ]:
labeled_barplot(df, "Brand", perc=True, n=10)

* Most of the cars in the data belong to Maruti or Hyundai.

In [ ]:
labeled_barplot(df, "Model", perc=True, n=10)

**Observations**

* Maruti Swift is the most common car up for resale.

* It is clear from the above charts that our dataset contains used cars from luxury as well as budget-friendly brands.

* We can create a new variable using this information. We can consider binning all our cars into the following 3 categories later:

    1. Budget-Friendly
    2. Mid Range
    3. Luxury Cars

In [ ]:
# dropping the redundant columns
df.drop(columns=["Kilometers_Driven", "Name", "mileage_unit"], inplace=True)

### Bivariate Distributions

**Let's check the correlation between numerical variables.**

In [ ]:
plt.figure(figsize=(15, 7))
sns.heatmap(df.corr(), annot=True, vmin=-1, vmax=1, fmt=".2f", cmap="Spectral")
plt.show()

**Observations**

* *Power* and *Engine* are important predictors of price, but they are also highly correlated to each other.
* We will have to work on imputing the missing values in *New_Price* because this is a very important feature in predicting the used car price accurately.

In [ ]:
sns.pairplot(data=df, hue="Fuel_Type")
plt.show()

**Observations**

Zooming into these plots gives us a lot of information.

* Contrary to intuition, *Kilometers_Driven* does not seem to have a relationship with the price.
* *Price* has a positive relationship with *Year*, i.e., the newer the car, the higher the price.
    - The temporal element of variation is captured in the year column.


* *S.No.* does not capture any information that we were hoping for.
* 2 seater cars are all luxury variants. Cars with 8-10 seats are exclusively mid to high range.
* Mileage does not seem to show much relationship with the price of used cars.
* Engine displacement and power of the car have a positive relationship with the price.
* *New_Price* and used car price are also positively correlated, which is expected.
* *Kilometers_Driven* has a peculiar relationship with the *Year* variable. Generally, the newer the car lesser the distance it has traveled, but this is not always true.
* CNG cars are conspicuous outliers when it comes to *Mileage*. The mileage of these cars is very high.
* The mileage and power of newer cars are increasing owing to advancements in technology.
* Mileage has a negative correlation with engine displacement and power. More powerful the engine, the more fuel it consumes in general.

**Let's check the variation in Price with some of the other variables.**

***Price* vs *Location***

In [ ]:
plt.figure(figsize=(12, 6))
sns.boxplot(x="Location", y="Price", data=df)
plt.show()

* The price of used cars has a large IQR in Coimbatore and Bangalore.

***Price* vs *Transmission***

In [ ]:
plt.figure(figsize=(12, 6))
sns.boxplot(x="Transmission", y="Price", data=df)
plt.show()

* The price is higher for used cars with automatic transmission.

***Price* vs *Year***

In [ ]:
plt.figure(figsize=(18, 6))
sns.boxplot(x="Year", y="Price", data=df)
plt.show()

* The price of used cars has increased over the years.

## Data Preprocessing (contd.)

### Missing Value Treatment

In [72]:
# let's check again for missing values
df.isnull().sum()

Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 46
Power                 175
Seats                  53
New_Price            6246
Price                1233
mileage_num             2
mileage_unit            2
engine_num             46
power_num             175
new_price_num        6246
Brand                   0
dtype: int64

<IPython.core.display.Javascript object>

**Let's look at a few rows where number of seats is missing.**

In [ ]:
df[df["Seats"].isnull()]

**We'll impute these missing values one-by-one by taking the median number of seats for the particular car using the Brand and Model name.**

In [ ]:
df.groupby(["Brand", "Model"], as_index=False)["Seats"].median()

In [ ]:
# imputing missing values in Seats
df["Seats"] = df.groupby(["Brand", "Model"])["Seats"].transform(
    lambda x: x.fillna(x.median())
)

In [ ]:
# Checking missing values in Seats
df[df["Seats"].isnull()]

* Maruti Estilo can accommodate 5 people.

In [ ]:
df["Seats"] = df["Seats"].fillna(5.0)

**We will use a similar method to fill missing values for the `Mileage`, `engine_num`, `power_num`, and `new_price_num` columns.**

In [ ]:
cols_list = ["mileage_num", "engine_num", "power_num", "new_price_num"]

for col in cols_list:
    df[col] = df.groupby(["Brand", "Model"])[col].transform(
        lambda x: x.fillna(x.median())
    )

df.isnull().sum()

* There are still some missing values in *Mileage*, *power_num*, and *new_price_num*.
* There are a few car brands and models in our dataset that do not contain the new price information at all.
* We'll impute these missing values one-by-one by taking the median number of seats for the particular car using the Brand name.

In [ ]:
cols_list = ["mileage_num", "power_num", "new_price_num"]

for col in cols_list:
    df[col] = df.groupby(["Brand"])[col].transform(lambda x: x.fillna(x.median()))

df.isnull().sum()

* There are still some missing values in *power_num* and *new_price_num*.
* We'll impute these missing values using the median of the data.

In [ ]:
cols_list = ["power_num", "new_price_num"]

for col in cols_list:
    df[col] = df[col].fillna(df[col].median())

df.isnull().sum()

- Missing values in all columns except *Price* and *price_log* are imputed.
- We will drop the rows where *Price* is missing.

In [ ]:
df = df[df["Price"].notna()]

## Linear Model Building

1. We want to predict the used car price, so we will use the normalized version *price_log* for modeling.

2. Before we proceed to build a model, we'll have to encode categorical features.

3. We'll split the data into train and test to be able to evaluate the model that we build on the train data.

4. We will build a Linear Regression model using the train data and then check it's performance.

In [ ]:
# defining the dependent and independent variables
X = df.drop(["Price", "price_log"], axis=1)
y = df["price_log"]

In [ ]:
# creating dummy variables
X = pd.get_dummies(
    X,
    columns=X.select_dtypes(include=["object", "category"]).columns.tolist(),
    drop_first=True,
)

X.head()

In [ ]:
# splitting the data in 70:30 ratio for train to test data

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

In [ ]:
print("Number of rows in train data =", x_train.shape[0])
print("Number of rows in test data =", x_test.shape[0])

In [ ]:
# fitting a linear model
lin_reg_model = LinearRegression()
lin_reg_model.fit(x_train, y_train)

In [ ]:
# let us check the coefficients and intercept of the model

coef_df = pd.DataFrame(
    np.append(lin_reg_model.coef_, lin_reg_model.intercept_),
    index=x_train.columns.tolist() + ["Intercept"],
    columns=["Coefficients"],
)
coef_df

### Model Performance Check

* We will check the model performance on the actual prices and not the log values.
* We will create a function that will convert the log prices to actual prices and then check the performance.
* We will be using metric functions defined in sklearn for RMSE, MAE, and $R^2$.
* We will define a function to calculate MAPE and adjusted $R^2$.
    - The mean absolute percentage error (MAPE) measures the accuracy of predictions as a percentage, and can be calculated as the average absolute percent error for each predicted value minus actual values divided by actual values. It works best if there are no extreme values in the data and none of the actual values are 0.

In [ ]:
# function to compute adjusted R-squared
def adj_r2_score(predictors, targets, predictions):
    r2 = r2_score(targets, predictions)
    n = predictors.shape[0]
    k = predictors.shape[1]
    return 1 - ((1 - r2) * (n - 1) / (n - k - 1))


# function to compute MAPE
def mape_score(targets, predictions):
    return np.mean(np.abs(targets - predictions) / targets) * 100


# function to compute different metrics to check performance of a regression model
def model_performance_regression(model, predictors, target):
    """
    Function to compute different metrics to check regression model performance

    model: regressor
    predictors: independent variables
    target: dependent variable
    """

    # predicting using the independent variables
    pred = model.predict(predictors)

    # computing the actual prices by using the exponential function
    target = np.exp(target)
    pred = np.exp(pred)

    r2 = r2_score(target, pred)  # to compute R-squared
    adjr2 = adj_r2_score(predictors, target, pred)  # to compute adjusted R-squared
    rmse = np.sqrt(mean_squared_error(target, pred))  # to compute RMSE
    mae = mean_absolute_error(target, pred)  # to compute MAE
    mape = mape_score(target, pred)  # to compute MAPE

    # creating a dataframe of metrics
    df_perf = pd.DataFrame(
        {
            "RMSE": rmse,
            "MAE": mae,
            "R-squared": r2,
            "Adj. R-squared": adjr2,
            "MAPE": mape,
        },
        index=[0],
    )

    return df_perf

In [ ]:
# Checking model performance on train set
print("Training Performance:")
lin_reg_model_perf_train = model_performance_regression(lin_reg_model, x_train, y_train)
lin_reg_model_perf_train

* Both the R-squared and Adjusted R squared of our model are very high. This is a clear indication that we have been able to create a very good model that is able to explain variance in the price of used cars up to 94%.

* The model is not an underfitting model.

* Let's do a quick performance check on the test data.

In [ ]:
# Checking model performance on test set
print("Test Performance:")
lin_reg_model_perf_test = model_performance_regression(lin_reg_model, x_test, y_test)
lin_reg_model_perf_test

* Root Mean Squared Error (RMSE) and Mean Absolute Error (MAE) of train and test data are close, which indicates that our model is not overfitting the train data.

* MAE indicates that our current model is able to predict used car prices within a mean error of 1.3 lakhs on test data.

* The units of both RMSE and MAE are the same, Lakhs in this case. But RMSE is greater than MAE because it penalizes the outliers more.

* MAPE of 12.98 on the test data indicates that the model can predict within ~13% of the used car price.

**Our current model is extremely complex. Let us bin the Brand and Model columns to make the dataset more manageable.**

In [ ]:
df.groupby(["Brand", "Model"])["new_price_num"].mean().sort_values(ascending=False)

**We will create a new variable `car_category` by binning the `new_price_num` column.**

In [ ]:
# creating a copy of the dataframe
df1 = df.copy()

# creating a new variable car_category
df1["car_category"] = pd.cut(
    x=df["new_price_num"],
    bins=[0, 15, 30, 50, 200],
    labels=["Budget_Friendly", "Mid-Range", "Luxury_Cars", "Ultra_luxury"],
)

df1.car_category.value_counts()

In [ ]:
# dropping the Brand and Model columns
df1.drop(columns=["Brand", "Model"], axis=1, inplace=True)

In [ ]:
# defining the dependent and independent variables
X = df1.drop(["Price", "price_log"], axis=1)
y = df1["price_log"]

# creating dummy variables
X = pd.get_dummies(
    X,
    columns=X.select_dtypes(include=["object", "category"]).columns.tolist(),
    drop_first=True,
)

# split the data into train and test
from sklearn.model_selection import train_test_split

x_train2, x_test2, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1
)

print("Number of rows in train data =", x_train2.shape[0])
print("Number of rows in test data =", x_test2.shape[0])

In [ ]:
lin_reg_model2 = LinearRegression()
lin_reg_model2.fit(x_train2, y_train)

In [ ]:
# Checking model performance on train set
print("Training Performance:")
lin_reg_model2_perf_train = model_performance_regression(
    lin_reg_model2, x_train2, y_train
)
lin_reg_model2_perf_train

In [ ]:
# Checking model performance on test set
print("Test Performance:")
lin_reg_model2_perf_test = model_performance_regression(lin_reg_model2, x_test2, y_test)
lin_reg_model2_perf_test

**Observations**

- The training performance has dropped drastically after binning the columns together.
- The test performance, however, remains close to the previous one.

### Forward Feature Selection

- Let us try using forward feature selection on the data before binning columns to check if we can get a good model performance using a lesser number of features.

**Note: Kindly do not run the code cells containing the Sequential Feature Selector implementation during the session, since that algorithm takes considerable time to run.**

In [ ]:
from mlxtend.feature_selection import SequentialFeatureSelector as SFS

reg = LinearRegression()

# Build step forward feature selection
sfs = SFS(
    reg,
    k_features=x_train.shape[1],
    forward=True,  # k_features denotes "Number of features to select"
    floating=False,
    scoring="r2",
    verbose=2,
    n_jobs=-1,  # this will ensure all CPU cores are being used for computation
    cv=5,
)

# Perform SFFS
sfs = sfs.fit(x_train, y_train)

In [ ]:
# to plot the performance with addition of each feature
from mlxtend.plotting import plot_sequential_feature_selection as plot_sfs

fig1 = plot_sfs(sfs.get_metric_dict(), kind="std_err", figsize=(15, 5))
plt.title("Sequential Forward Selection (w. StdErr)")
plt.xticks(
    np.arange(0, 264, 10), np.arange(0, 264, 10), rotation=90
)  # to make the tick marks readable
plt.show()

* The model performance appears to have become constant somewhere around 118 features and then starts falling around 239 features.
* The improvement in performance from 85 to 118 features is not that high either.
* We will take 85 as the number of features to consider.

In [ ]:
reg = LinearRegression()

# # Build step forward feature selection
sfs = SFS(
    reg,
    k_features=85,
    forward=True,
    floating=False,
    scoring="r2",
    verbose=2,
    cv=5,
    n_jobs=-1,
)

# Perform SFFS
sfs = sfs.fit(x_train, y_train)

In [ ]:
# important features
feat_cols = list(sfs.k_feature_idx_)
print(feat_cols)

In [ ]:
x_train.columns[feat_cols]

**Now we will fit a sklearn model using these features only.**

In [ ]:
x_train3 = x_train[x_train.columns[feat_cols]]

In [ ]:
# Creating new x_test with the same 20 variables that we selected for x_train
x_test3 = x_test[x_train3.columns]

In [ ]:
# Fitting linear model
lin_reg_model3 = LinearRegression()
lin_reg_model3.fit(x_train3, y_train)

# let us check the coefficients and intercept of the model

coef_df = pd.DataFrame(
    np.append(lin_reg_model3.coef_, lin_reg_model3.intercept_.flatten()),
    index=x_train3.columns.tolist() + ["Intercept"],
    columns=["Coefficients"],
)
coef_df

In [ ]:
# model performance on train set
print("Training Performance:")
lin_reg_model3_perf_train = model_performance_regression(
    lin_reg_model3, x_train3, y_train
)
lin_reg_model3_perf_train

In [ ]:
# model performance on test set
print("Test Performance:")
lin_reg_model3_perf_test = model_performance_regression(lin_reg_model3, x_test3, y_test)
lin_reg_model3_perf_test

* We have been able to explain ~90% of the variation in the data using less than one-third of the number of features used in the previous model.
* Let's compare the two models we built.

In [ ]:
# training performance comparison

models_train_comp_df = pd.concat(
    [
        lin_reg_model_perf_train.T,
        lin_reg_model2_perf_train.T,
        lin_reg_model3_perf_train.T,
    ],
    axis=1,
)

models_train_comp_df.columns = [
    "Linear Regression sklearn",
    "Linear Regression sklearn (binned columns)",
    "Linear Regression sklearn (SFS features)",
]

print("Training performance comparison:")
models_train_comp_df

In [ ]:
# test performance comparison

models_test_comp_df = pd.concat(
    [lin_reg_model_perf_test.T, lin_reg_model2_perf_test.T, lin_reg_model3_perf_test.T],
    axis=1,
)

models_test_comp_df.columns = [
    "Linear Regression sklearn",
    "Linear Regression sklearn (binned columns)",
    "Linear Regression sklearn (SFS features)",
]

print("Test performance comparison:")
models_test_comp_df

**We will consider the model with SFS features (*lin_reg_model3*) as our final model.**

## Observations

1. With our linear regression model *lin_reg_model3*, we have been able to capture ~90% of the variation in the data.


2. MAE indicates that our current model is able to predict used car prices within a mean error of approx. 1.4 lakhs on the test data, and the MAPE indicates that the model can predict within ~14% of the used car price.


3. Factors like the year of manufacture, numbers of seats, maximum power of the engine, few locations, few brands, etc. tend to increase the price of the used car.


4. Factors like distance covered in unit fuel, log of the number of kilometers drive, few locations, few brands, etc. tend to decrease the price of the used car.

## Business Insights and Recommendations

* Cars with a lesser number of kilometers driven should be preferred. 


* Some markets tend to have higher prices. Cars4U should focus more on these markets, and set up offices in these areas if needed.


* We will have to analyze the cost side of things before we can talk about profitability in the business. We should gather data regarding that.


* The next step post that would be to cluster different sets of data and see if we should make multiple models for different locations/car types.

## Add-on: Analysing predictions where we were way off the mark

In [ ]:
# Extracting the rows from original data frame df where indexes are same as the training data
original_df = df[df.index.isin(x_train2.index.values)].copy()

# Extracting predicted values and residuals from the final model
fitted_values = lin_reg_model2.predict(x_train2)
residuals = fitted_values - y_train

# Add new columns for predicted values
original_df["Predicted price_log "] = fitted_values
original_df["Predicted Price"] = np.exp(fitted_values)
original_df["residuals"] = residuals
original_df["Abs_residuals"] = np.exp(residuals)
original_df["Difference in Lakhs"] = np.abs(
    original_df["Price"] - original_df["Predicted Price"]
)

# Let us look at the top 20 predictions where our model made highest extimation errors (on train data)
original_df.sort_values(by=["Difference in Lakhs"], ascending=False).head(20)

**Observations**

* A 2017 Land Rover, whose new model sells at 230 Lakhs and the used version sold at 160 Lakhs was predicted to be sold at < 3 Lakhs. It is not apparent after looking at numerical predictors, why our model predicted such low value here. This could be because many other land rovers in our data seem to have sold at lower prices.

* Another entry in the list here is a Lamborghini Gallardo that was sold at 120 Lakhs but our model predicted the price around 3 lakhs. This is a huge error by the model. However, there might be a data entry error here as the price of a new Gallardo is set at around 11 Lakhs, which is less than the selling price of a used Gallardo.

* There are a few instances where the model predicts lesser than the actual selling price. These could be a cause for concern. The model predicting lesser than potential selling price is not good for business.


Let us quickly visualize some of these observations. 

In [ ]:
sns.scatterplot(
    data=original_df,
    x="Difference in Lakhs",
    y="Predicted Price",
    hue=original_df["Fuel_Type"],
)
plt.show()

- Our model predicts that the resale value of diesel cars is higher compared to petrol cars.

## Add-on: Advanced Data Preprocessing

In [ ]:
df = data.copy()

**1. `Mileage`**

* We have car mileage in two units as per the data dictionary - kmpl and km/kg.

* After quick research on the internet, it is clear that these 2 units are used for cars of 2 different fuel types.

    * kmpl (kilometers per litre) is used for petrol and diesel cars.
    * km/kg (kilometers per kg) is used for CNG and LPG-based engines.

* We have the variable `Fuel_type` in our data. Let us check if these observations hold true in our data also.

In [ ]:
# Create 2 new columns after splitting the mileage values.
mileage_num = []
mileage_unit = []

for observation in df["Mileage"]:
    if isinstance(observation, str):
        if (
            observation.split(" ")[0]
            .replace(".", "", 1)
            .isdigit()  # first element should be numeric
            and " " in observation  # space between numeric and unit
            and (
                observation.split(" ")[1]
                == "kmpl"  # units are limited to "kmpl" and "km/kg"
                or observation.split(" ")[1] == "km/kg"
            )
        ):
            mileage_num.append(float(observation.split(" ")[0]))
            mileage_unit.append(observation.split(" ")[1])
        else:
            # To detect if there are any observations in the column that do not follow
            # the expected format [number + ' ' + 'kmpl' or 'km/kg']
            print(
                "The data needs further processing. All values are not similar ",
                observation,
            )
    else:
        # If there are any missing values in the mileage column,
        # we add corresponding missing values to the 2 new columns
        mileage_num.append(np.nan)
        mileage_unit.append(np.nan)

* No print output from the function above. The values are all in the expected format or NaNs.

In [ ]:
# Add the new columns to the data

df["mileage_num"] = mileage_num
df["mileage_unit"] = mileage_unit

# Checking the new dataframe
df.head(5)

In [ ]:
# Let us check if the units correspond to the fuel types
df.groupby(by=["Fuel_Type", "mileage_unit"]).size()

* As expected, km/kg is for CNG/LPG cars and kmpl is for Petrol and Diesel cars.

**2. `Engine`**

* As per the data dictionary, the `Engine` column indicates the displacement volume of the engine in CC.

* We will make sure that all the observations follow the same format - [numeric + " " + "CC"] and create a new numeric column from this column.

* This time, let's use a regex to make all the necessary checks.

In [ ]:
# re module provides support for regular expressions
import re

# Create a new column after splitting the engine values.
engine_num = []

# Regex for numeric + " " + "CC"  format
regex_engine = "^\d+(\.\d+)? CC$"

for observation in df["Engine"]:
    if isinstance(observation, str):
        if re.match(regex_engine, observation):
            engine_num.append(float(observation.split(" ")[0]))
        else:
            # To detect if there are any observations in the column that do not follow [numeric + " " + "CC"]  format
            print(
                "The data needs furthur processing. All values are not similar ",
                observation,
            )
    else:
        # If there are any missing values in the engine column, we add missing values to the new column
        engine_num.append(np.nan)

* No print output from the function above. The values are all in the same format - [numeric + " " + "CC"] or NaNs.

In [ ]:
# Add the new column to the data
df["engine_num"] = engine_num

# Checking the new dataframe
df.head()

**3. `Power`**

* As per the data dictionary, the `Power` column indicates the maximum power of the engine in bhp.

* We will make sure that all the observations follow the same format - [numeric + " " + "bhp"] and create a new numeric column from this column like we did for the `Engine` column.

In [ ]:
# Create a new column after splitting the power values.
power_num = []

# Regex for numeric + " " + "bhp"  format
regex_power = "^\d+(\.\d+)? bhp$"

for observation in df["Power"]:
    if isinstance(observation, str):
        if re.match(regex_power, observation):
            power_num.append(float(observation.split(" ")[0]))
        else:
            # To detect if there are any observations in the column that do not follow [numeric + " " + "bhp"]  format
            # that we see in the sample output
            print(
                "The data needs furthur processing. All values are not similar ",
                observation,
            )
    else:
        # If there are any missing values in the power column, we add missing values to the new column
        power_num.append(np.nan)

* No print output from the function above. The values are all in the same format - [numeric + " " + "bhp"] or NaNs.

In [ ]:
# Add the new column to the data
df["power_num"] = power_num

# Checking the new dataframe
df.head()

**4. `New_Price`**

* We know that `New_Price` is the price of a new car of the same model in INR Lakhs.(1 Lakh INR = 100,000 INR)

* This column clearly has a lot of missing values. We will impute the missing values later. For now we will only extract the numeric values from this column.

In [ ]:
# Create a new column after splitting the New_Price values.
new_price_num = []

# Regex for numeric + " " + "Lakh"  format
regex_power = "^\d+(\.\d+)? Lakh$"

for observation in df["New_Price"]:
    if isinstance(observation, str):
        if re.match(regex_power, observation):
            new_price_num.append(float(observation.split(" ")[0]))
        else:
            # To detect if there are any observations in the column that do not follow [numeric + " " + "Lakh"]  format
            # that we see in the sample output
            print(
                "The data needs furthur processing. All values are not similar ",
                observation,
            )
    else:
        # If there are any missing values in the New_Price column, we add missing values to the new column
        new_price_num.append(np.nan)

* Not all values are in Lakhs.

* There are a few observations that are in Crores as well

* Let us convert these to lakhs using the conversion rate 1 Cr = 100 Lakhs.

In [ ]:
new_price_num = []

for observation in df["New_Price"]:
    if isinstance(observation, str):
        if re.match(regex_power, observation):
            new_price_num.append(float(observation.split(" ")[0]))
        else:
            # Converting values in Crore to lakhs
            new_price_num.append(float(observation.split(" ")[0]) * 100)
    else:
        # If there are any missing values in the New_Price column, we add missing values to the new column
        new_price_num.append(np.nan)

# Add the new column to the data
df["new_price_num"] = new_price_num

# Checking the new dataframe
df.head(5)

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Note:-This-is-a-sample-solution-for-the-project.-Projects-will-NOT-be-graded-on-the-basis-of-how-well-the-submission-matches-this-sample-solution.-Projects-will-be-graded-on-the-basis-of-the-rubric-only." data-toc-modified-id="Note:-This-is-a-sample-solution-for-the-project.-Projects-will-NOT-be-graded-on-the-basis-of-how-well-the-submission-matches-this-sample-solution.-Projects-will-be-graded-on-the-basis-of-the-rubric-only.-1">Note: This is a sample solution for the project. Projects will NOT be graded on the basis of how well the submission matches this sample solution. Projects will be graded on the basis of the rubric only.</a></span></li><li><span><a href="#Context" data-toc-modified-id="Context-2">Context</a></span></li><li><span><a href="#Objective" data-toc-modified-id="Objective-3">Objective</a></span></li><li><span><a href="#Data-Description" data-toc-modified-id="Data-Description-4">Data Description</a></span></li><li><span><a href="#Let's-start-coding!" data-toc-modified-id="Let's-start-coding!-5">Let's start coding!</a></span><ul class="toc-item"><li><span><a href="#Importing-necessary-libraries" data-toc-modified-id="Importing-necessary-libraries-5.1">Importing necessary libraries</a></span></li></ul></li><li><span><a href="#Data-Preprocessing" data-toc-modified-id="Data-Preprocessing-6">Data Preprocessing</a></span><ul class="toc-item"><li><span><a href="#Processing-columns" data-toc-modified-id="Processing-columns-6.1">Processing columns</a></span></li><li><span><a href="#Feature-Engineering" data-toc-modified-id="Feature-Engineering-6.2">Feature Engineering</a></span><ul class="toc-item"><li><span><a href="#1.-Car-Brand-Name" data-toc-modified-id="1.-Car-Brand-Name-6.2.1">1. Car Brand Name</a></span></li><li><span><a href="#2.-Car-Model-Name" data-toc-modified-id="2.-Car-Model-Name-6.2.2">2. Car Model Name</a></span></li></ul></li></ul></li><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-7">Exploratory Data Analysis</a></span><ul class="toc-item"><li><span><a href="#Univariate-Analysis" data-toc-modified-id="Univariate-Analysis-7.1">Univariate Analysis</a></span></li><li><span><a href="#Kilometers_Driven" data-toc-modified-id="Kilometers_Driven-7.2"><em>Kilometers_Driven</em></a></span></li><li><span><a href="#Bivariate-Distributions" data-toc-modified-id="Bivariate-Distributions-7.3">Bivariate Distributions</a></span></li></ul></li><li><span><a href="#Data-Preprocessing-(contd.)" data-toc-modified-id="Data-Preprocessing-(contd.)-8">Data Preprocessing (contd.)</a></span><ul class="toc-item"><li><span><a href="#Missing-Value-Treatment" data-toc-modified-id="Missing-Value-Treatment-8.1">Missing Value Treatment</a></span></li></ul></li><li><span><a href="#Linear-Model-Building" data-toc-modified-id="Linear-Model-Building-9">Linear Model Building</a></span><ul class="toc-item"><li><span><a href="#Model-Performance-Check" data-toc-modified-id="Model-Performance-Check-9.1">Model Performance Check</a></span></li><li><span><a href="#Forward-Feature-Selection" data-toc-modified-id="Forward-Feature-Selection-9.2">Forward Feature Selection</a></span></li></ul></li><li><span><a href="#Observations" data-toc-modified-id="Observations-10">Observations</a></span></li><li><span><a href="#Business-Insights-and-Recommendations" data-toc-modified-id="Business-Insights-and-Recommendations-11">Business Insights and Recommendations</a></span></li><li><span><a href="#Add-on:-Analysing-predictions-where-we-were-way-off-the-mark" data-toc-modified-id="Add-on:-Analysing-predictions-where-we-were-way-off-the-mark-12">Add-on: Analysing predictions where we were way off the mark</a></span></li><li><span><a href="#Add-on:-Advanced-Data-Preprocessing" data-toc-modified-id="Add-on:-Advanced-Data-Preprocessing-13">Add-on: Advanced Data Preprocessing</a></span></li></ul></div>